contracts/contract_call.go

```diff
@@ -0,0 +1,112 @@
+package contracts
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/accounts/abi"
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/common/hexutil"
+	"github.com/celo-org/celo-blockchain/core/vm"
+	"github.com/celo-org/celo-blockchain/log"
+)
+
+// Method represents a contract's method
+type Method struct {
+	abi    *abi.ABI
+	method string
+	maxGas uint64
+}
+
+// NewMethod creates a new contract message
+func NewMethod(abi *abi.ABI, method string, maxGas uint64) Method {
+	return Method{
+		abi:    abi,
+		method: method,
+		maxGas: maxGas,
+	}
+}
+
+func (am Method) Bind(contractAddress common.Address) *BoundMethod {
+	return &BoundMethod{
+		Method:         am,
+		resolveAddress: noopResolver(contractAddress),
+	}
+}
+
+// encodeCall will encodes the msg into []byte format for EVM consumption
+func (am Method) encodeCall(args ...interface{}) ([]byte, error) {
+	return am.abi.Pack(am.method, args...)
+}
+
+// decodeResult will decode the result of msg execution into the result parameter
+func (am Method) decodeResult(result interface{}, output []byte) error {
+	if result == nil {
+		return nil
+	}
+	return am.abi.Unpack(result, am.method, output)
+}
+
+func NewBoundMethod(contractAddress common.Address, abi *abi.ABI, methodName string, maxGas uint64) *BoundMethod {
+	return NewMethod(abi, methodName, maxGas).Bind(contractAddress)
+}
+
+func NewRegisteredContractMethod(registryId common.Hash, abi *abi.ABI, methodName string, maxGas uint64) *BoundMethod {
+	return &BoundMethod{
+		Method: NewMethod(abi, methodName, maxGas),
+		resolveAddress: func(vmRunner vm.EVMRunner) (common.Address, error) {
+			return resolveAddressForCall(vmRunner, registryId, methodName)
+		},
+	}
+}
+
+type BoundMethod struct {
```

Comment:
I was trying to add some suggested documentation for BoundMethod but I found that I wasn't able to easily express what it is. I think that the naming is not quite right here, bound implies locked to something and makes sense in the context of the Bind method.

```go
func (am Method) Bind(contractAddress common.Address) *BoundMethod {
```

Because `Bind` is binding the Method to a specific address, but the type BoundMethod is not actually bound to a specific address in fact it has a function to resolve the address when invoked, meaning that the address can change during the lifetime of the object, which is incongruent with the concept of bound.

Also given that Method doesn't actually provide any functionality to be called I think it would be clearer to rename Method →MethodTemplate/MethodSignature and BoundMethod→Method.

Reply:
Interesting point about `maxGas`, if you are saying that it does not have anything to do with the signature (which I agree with) then I think it must relate to the implementation, which means that it should be a field on `BoundMethod` because that can change based on the address that a method is bound to.